src/logging.cpp

```diff
@@ -121,7 +142,11 @@ bool BCLog::Logger::DisableCategory(const std::string& str)
 
 bool BCLog::Logger::WillLogCategory(BCLog::LogFlags category) const
 {
-    return (m_categories.load(std::memory_order_relaxed) & category) != 0;
+    if (Assume(category < NONE)) {
+        StdLockGuard scoped_lock(m_cs);
+        return m_categories.test(category);
+    }
+    return false;
```

Comment:
Not 100% sure about this. Feels a bit brittle to use `Assume` here. Previously it was fine to pass `ALL` or `NONE` here - we could keep that behaviour?

Reply:
Currently, neither `NONE` nor `ALL` is passed to this function. The only caller to this function is `LogAcceptCategory()`, which is called by `LogPrint()` and similar, and only actual logging categories (for example, `BCLog::NET`) are passed to those, never `NONE` or `ALL`. I don't think passing those makes sense. `NONE` would mean don't log at all (so why is the `LogPrint` statement there to begin with?), and `ALL` would be just the same as `LogPrintf()`. We could make those work, I guess. But, as I said, they're not used today.

But here I would like to not use `NONE`; better to use `m_categories.size()`.

I think the `Assume()` is nice to keep, because we would like to know if someone calls `LogPrint(BCLog::NONE, ...)` (or similar).

Reply:
Yeah they're not currently being used, but I think (unnecessarily) bounding which elements of an enum we can accept is confusing at best since that exception is not immediately obvious from the function signature. It can also make maintenance more burdensome. And even _if_ we don't see any use cases (and I think, for example, it's not unreasonable for a future callsite to check if  _any_ logging is happening at the moment), I still think we shouldn't necessarily prevent `ALL` or `NONE` from being used.

As an alternative solution, I think having a single function to convert `LogFlags` into `LogFlagsBitSet` simplifies the code and allows callsites to accept any kind of LogFlags:

<details>
<summary>git diff on 0dbe9b487</summary>

```diff
diff --git a/src/logging.cpp b/src/logging.cpp
index 970f99c12..6b24f7fe2 100644
--- a/src/logging.cpp
+++ b/src/logging.cpp
@@ -94,19 +94,21 @@ void BCLog::Logger::DisconnectTestLogger()
     m_print_callbacks.clear();
 }
 
+/** Convert LogFlags into their equivalent bitset for generalized bitwise comparison */
+BCLog::LogFlagsBitset FlagToBitset(BCLog::LogFlags flag){
+    BCLog::LogFlagsBitset flag_bitset;
+    switch(flag) {
+        case BCLog::LogFlags::NONE: break;
+        case BCLog::LogFlags::ALL: flag_bitset.set(); break;
+        default: flag_bitset.set(flag); break;
+    }
+    return flag_bitset;
+}
+
 void BCLog::Logger::EnableCategory(BCLog::LogFlags flag)
 {
     StdLockGuard scoped_lock(m_cs);
-    switch (flag) {
-        case BCLog::NONE:
-            return;
-        case BCLog::ALL:
-            // with no argument, set() sets all the bits to true
-            m_categories.set();
-            return;
-        default: break;
-    }
-    if (Assume(flag < NONE)) m_categories.set(flag);
+    m_categories |= FlagToBitset(flag);
 }
 
 bool BCLog::Logger::EnableCategory(const std::string& str)
@@ -120,16 +122,7 @@ bool BCLog::Logger::EnableCategory(const std::string& str)
 void BCLog::Logger::DisableCategory(BCLog::LogFlags flag)
 {
     StdLockGuard scoped_lock(m_cs);
-    switch (flag) {
-        case BCLog::NONE:
-            return;
-        case BCLog::ALL:
-            // set all the category flags to false
-            m_categories = 0;
-            return;
-        default: break;
-    }
-    if (Assume(flag < NONE)) m_categories.reset(flag);
+    m_categories &= ~FlagToBitset(flag);
 }
 
 bool BCLog::Logger::DisableCategory(const std::string& str)
@@ -139,14 +132,10 @@ bool BCLog::Logger::DisableCategory(const std::string& str)
     DisableCategory(flag);
     return true;
 }
-
 bool BCLog::Logger::WillLogCategory(BCLog::LogFlags category) const
 {
-    if (Assume(category < NONE)) {
-        StdLockGuard scoped_lock(m_cs);
-        return m_categories.test(category);
-    }
-    return false;
+    StdLockGuard scoped_lock(m_cs);
+    return (m_categories & FlagToBitset(category)) != 0;
 }
 
 bool BCLog::Logger::WillLogCategoryLevel(BCLog::LogFlags category, BCLog::Level level) const
```
</details>

_(Note: at this point, we also almost just as well might have `m_categories` be a `uint64_t` instead of `std::bitset`, but I'm not very familar with the trade-offs between `std::bitset` and `uint64_t` - it seems bitsets can be a bit safer to use?)_

Reply:
I guess in more generalized terms: I think in this approach it's nice that we have a single place (`FlagToBitset()`) where we define what each LogFlag means, as opposed to having to redefine that in every function.

Reply:
> I still think we shouldn't necessarily prevent ALL or NONE from being used.

I like this idea, I just force-pushed an update (40e1923e856661fdf68cf783ed9c6d1931dfbdf3) to allow those to have the obvious meanings when specified by `LogPrint()` (i.e. `WillLogCategory()`).

As for `FlagToBitset()`, that's very interesting, but it has the downside that evaluating `m_categories & FlagToBitset(category)` (in `WillLogCategory()`) requires (in effect) checking all of the categories (bits), whereas `m_categories.test(category)` checks only one and so _feels_ more efficient (although it may not really be, depending on the implementation). Also, `FlagToBitset()` makes an actual function call to the constructor (I verified this using https://godbolt.org/), and there may also be a call from `WillLogCategory()` to `FlagToBitset()` (if the compiler doesn't inline it). Since these happen for testing a category, which is performance-critical, perhaps they should be avoided.

> we also almost just as well might have m_categories be a uint64_t ...

One nice advantage of using `std::bitset` is that the `BCLog::LogFlags` categories are a simple `enum`, rather than being defined with a manually-selected shift constant, like `1 << 28`. Also, this seems like a perfect use case for `std::bitset` and is just more modern than low-level bit manipulation. So I think I still like `std::bitset` here.